backend/agentd/session.go

```diff
@@ -214,7 +229,48 @@ func (s *Session) sender() {
 	for {
 		var msg *transport.Message
 		select {
-		case msg = <-s.sendq:
+		case e := <-s.entityConfig.updatesChannel:
+			watchEvent, ok := e.(*store.WatchEventEntityConfig)
+			if !ok {
+				logger.Errorf("session received unexpected struct: %T", e)
+				continue
+			}
+
+			if watchEvent.Entity == nil {
+				logger.Error("session received nil entity in watch event")
+				continue
+			}
+
+			logger.WithFields(logrus.Fields{
+				"action":    watchEvent.Action.String(),
+				"entity":    watchEvent.Entity.Metadata.Name,
+				"namespace": watchEvent.Entity.Metadata.Namespace,
+			}).Debug("entity update received")
+
+			// Handle the delete and unknown watch events
+			switch watchEvent.Action {
+			case store.WatchDelete:
+				// The entity was deleted, we should sever the connection to the agent
+				// so it can register back
+				s.cancel()
+				continue
+			case store.WatchUnknown:
+				logger.Error("session received unknown watch event")
+				continue
+			}
+
+			// Enforce the entity class to agent
+			if watchEvent.Entity.EntityClass != corev2.EntityAgentClass {
```

Comment:
In this case, we also need to update the entity in the store, so that it no longer has a non-agent class.

Reply:
Should we avoid publishing the entity update in that case and simply publish the watch event that will be result from the store update?

Reply:
Hm, good question. It's kind of a corner case. I don't think it particularly matters, but we could save some time if we avoid publishing, and just wait for the watch event to come in that will be triggered by updating the store.

Reply:
Just pushed a new commit that update the entity config when the class is misconfigured!